libcloud/compute/drivers/ecs.py

```diff
@@ -0,0 +1,1390 @@
+            return None
+        return self._get_extra_dict(
+            eips[0], RESOURCE_EXTRA_ATTRIBUTES_MAP['eip_address_associate'])
+
+    def _get_operation_locks(self, instance):
+        locks = findall(instance, xpath='OperationLocks',
+                        namespace=self.namespace)
+        if len(locks) <= 0:
+            return None
+        return self._get_extra_dict(
+            locks[0], RESOURCE_EXTRA_ATTRIBUTES_MAP['operation_locks'])
+
+    def _wait_until_state(self, nodes, state, wait_period=3, timeout=600):
+        """
+        Block until the provided nodes are in the desired state.
+        :param nodes: List of nodes to wait for
+        :type nodes: ``list`` of :class:`.Node`
+        :param state: desired state
+        :type state: ``NodeState``
+        :param wait_period: How many seconds to wait between each loop
+                            iteration. (default is 3)
+        :type wait_period: ``int``
+        :param timeout: How many seconds to wait before giving up.
+                        (default is 600)
+        :type timeout: ``int``
+        :return: if the nodes are in the desired state.
+        :rtype: ``bool``
+        """
+        start = time.time()
+        end = start + timeout
+        node_ids = [node.id for node in nodes]
+
+        while(time.time() < end):
+            matched_nodes = self.list_nodes(ex_node_ids=node_ids)
+            if len(matched_nodes) > len(node_ids):
+                found_ids = [node.id for node in matched_nodes]
+                msg = ('found multiple nodes with same ids, '
+                       'desired ids: %(ids)s, found ids: %(found_ids)s' %
+                       {'ids': node_ids, 'found_ids': found_ids})
+                raise LibcloudError(value=msg, driver=self)
+            desired_nodes = [node for node in matched_nodes
+                             if node.state == state]
+
+            if len(desired_nodes) == len(node_ids):
+                return True
+            else:
+                time.sleep(wait_period)
+                continue
+
+        raise LibcloudError(value='Timed out after %s seconds' % (timeout),
+                            driver=self)
+
+    def _to_volume(self, element):
+        _id = findtext(element, 'DiskId', namespace=self.namespace)
+        name = findtext(element, 'DiskName', namespace=self.namespace)
+        size = int(findtext(element, 'Size', namespace=self.namespace))
+        status_str = findtext(element, 'Status', namespace=self.namespace)
+        status = self.VOLUME_STATE_MAPPING.get(status_str,
+                                               StorageVolumeState.UNKNOWN)
+
+        extra = self._get_extra_dict(element,
+                                     RESOURCE_EXTRA_ATTRIBUTES_MAP['volume'])
+        extra['operation_locks'] = self._get_operation_locks(element)
+        return StorageVolume(_id, name, size, self, state=status, extra=extra)
+
+    def _list_to_json_array(self, value):
```

Comment:
I should use it all the time except `list_images` method, in which the API need a comma separated string instead of a json array string.